File: Kaskus/GridViewPaging/GridViewPagingVb/Pages/PageList.vb
Imports System
Imports System.Collections
Imports System.ComponentModel
Imports System.Linq

Namespace Pages

    Friend Class PageList
        Implements IListSource

        Private ReadOnly _pageSize As Int32

        Public Sub New(pageSize As Int32, totalrecord As Int64)
            Me.TotalRecord = totalrecord
            _pageSize = pageSize
        End Sub

        Public Property TotalRecord() As Int64

        Public ReadOnly Property ContainsListCollection() As [Boolean] Implements IListSource.ContainsListCollection
            Get
                Return (GetList().Count >= 1)
            End Get
        End Property

        Public Function GetList() As IList Implements IListSource.GetList
            Return Enumerable.Range(0, CType(Math.Ceiling(CType(TotalRecord, [Double]) / _pageSize), Int32)).Select(Function(pageidx) New ItemPage(_pageSize, pageidx * _pageSize)).ToList()
        End Function

        Public NotInheritable Class ItemPage

            Public ReadOnly PageSize As Int32

            Public ReadOnly CurrentPage As Int32

            Public Sub New(limit As Int32, offset As Int32)
                PageSize = limit
                CurrentPage = offset
            End Sub
        End Class
    End Class
End Namespace

File: Kaskus/GridViewPaging/GridViewPagingCs/Pages/PageList.cs
namespace Dede.App.GridViewPaging.Pages
{
    using System;
    using System.Collections;
    using System.ComponentModel;
    using System.Linq;

    internal class PageList
        : IListSource
    {
        private readonly Int32 pageSize;

        public PageList(
            Int32 pageSize,
            Int64 totalrecord)
        {
            this.TotalRecord = totalrecord;
            this.pageSize    = pageSize;
        }

        public Int64 TotalRecord { get; set; }

        public Boolean ContainsListCollection { get; protected set; }

        public IList GetList()
        {
            return Enumerable
                .Range(0, (Int32)Math.Ceiling((Double)this.TotalRecord / this.pageSize))
                .Select(pageidx => new ItemPage(this.pageSize, pageidx * this.pageSize))
                .ToList();
        }

        public sealed class ItemPage
        {
            public readonly Int32 PageSize;

            public readonly Int32 CurrentPage;

            public ItemPage(
                Int32 limit,
                Int32 offset)
            {
                this.PageSize    = limit;
                this.CurrentPage = offset;
            }
        }
    }
}